### PR TITLE
Enable Blob module in UWP

### PR DESCRIPTION
**Pull request by JunielKatarn**
### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
Blob support should be available in both Desktop and Universal variants of React Native Windows.

### What
- Add Shared project to ReactUWPTestApp solution
- Have OInstance instantiate HTTP, Blob and FileReader modules **only** on Windows Desktop.
- Add Blob and FileReader modules to ReactInstanceImple core modules.

## Testing
Verified Blob module gets instantiated and JavaScript error `Invalid response for blob: ${this._response}` no longer happens.

_Optional_: Describe the tests that you ran locally to verify your changes.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/10187)

### CODE REVIEW

File: packages/e2e-test-app/windows/RNTesterApp/RNTesterApp.csproj [hunk 1]
```diff
@@ -145,7 +145,7 @@
     <Import Project="$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CSharpApp.props" Condition="Exists('$(ReactNativeWindowsDir)\PropertySheets\External\Microsoft.ReactNative.Uwp.CSharpApp.props')" />
   </ImportGroup>
   <ItemGroup>
-  <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
+    <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
```
Comment (JunielKatarn) [inline packages/e2e-test-app/windows/RNTesterApp/RNTesterApp.csproj:148]: Auto-formatting by Visual Studio.

File: packages/e2e-test-app/windows/ReactUWPTestApp.sln [hunk 2]
```diff
@@ -42,20 +42,16 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactNativePicker", "..\..\
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ReactNativeXaml", "..\..\..\node_modules\react-native-xaml\windows\ReactNativeXaml\ReactNativeXaml.vcxproj", "{0FF7027A-222C-4FFB-8F17-91D18BBAF7A8}"
 EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "Microsoft.ReactNative.Shared", "..\..\..\vnext\Shared\Shared.vcxitems", "{2049DBE9-8D13-42C9-AE4B-413AE38FFFD0}"
```
Comment (JunielKatarn) [inline packages/e2e-test-app/windows/ReactUWPTestApp.sln:45]: For reviewers:

Why was project `Microsoft.ReactNative.Shared` not included in the this solution?

File: vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp [hunk 3]
```diff
@@ -50,11 +51,24 @@ std::vector<facebook::react::NativeModuleDescription> GetCoreModules(
...
+        [props = context->Properties()]() { return Microsoft::React::CreateFileReaderModule(props); },
+        jsMessageQueue);
+  }
+
   modules.emplace_back(
       "Timing",
       [batchingUIMessageQueue]() { return facebook::react::CreateTimingModule(batchingUIMessageQueue); },
       batchingUIMessageQueue);
 
+  // TODO: For reviewers - Why does this factory move the context object?
```
Comment (JunielKatarn) [inline vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:71]: For reviewers.

File: vnext/Shared/OInstance.cpp [hunk 4]
```diff
@@ -545,13 +545,15 @@ std::vector<std::unique_ptr<NativeModule>> InstanceImpl::GetDefaultNativeModules
   std::vector<std::unique_ptr<NativeModule>> modules;
   auto transitionalProps{ReactPropertyBagHelper::CreatePropertyBag()};
 
+#if (defined(_MSC_VER) && !defined(WINRT))
```
Comment (asklar) [inline vnext/Shared/OInstance.cpp:548]: can you clarify this bit? why is the http module not available for uwp? (e.g. is it replaced by blob module?)

File: vnext/Shared/OInstance.cpp [hunk 5]
```diff
@@ -620,7 +622,9 @@ std::vector<std::unique_ptr<NativeModule>> InstanceImpl::GetDefaultNativeModules
       nativeQueue));
 
   // #10036 - Blob module not supported in UWP. Need to define property bag lifetime and onwership.
```
Comment (asklar) [inline vnext/Shared/OInstance.cpp:624]: this comment is no longer true? can remove the ifdef now?

**Review by asklar — CHANGES_REQUESTED** [review]

File: vnext/Shared/OInstance.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:548]: The HTTP was being instantiated twice in MSRN (last one wins, though).
For a few reasons, including the fact that the Desktop variant does not have access to the property bag, ( @acoates-ms should know more details), Desktop needs to instantiate some modules here (HTTP, WebSocket, Timer...) for them to be available to the consuming app.

These "core" modules are instantiated somewhere else in MSRN:
CoreNativeModules.cpp.

In summary, this part of the change:
- Removes duplicate instantiations of core modules
- Ensures modules that require a property bag to communicate (HTTP, Blob, FileReader, WebSocket) use the appropriate bag instance (Desktop: OInstance local object. Universal: context's member).

File: vnext/Shared/OInstance.cpp (hunk 5 above)
Comment (JunielKatarn) [reply vnext/Shared/OInstance.cpp:624]: The comment is no longer accurate. I'll update, but the ifdef should still be there. Otherwise, we'll create multiple instances of the same module. Not a functional problem, but an unnecessary use of extra memory.

File: packages/e2e-test-app/windows/ReactUWPTestApp.sln [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29102.190
+# Visual Studio Version 17
+VisualStudioVersion = 17.2.32602.215
```
[the remark this answers is not in the document]
Comment (JunielKatarn) [reply packages/e2e-test-app/windows/ReactUWPTestApp.sln:4]: It sure will not. I'll double check right now, but I recall making this experiment from VS 2013 to 2015.

In any case, this was don by the IDE and the actual change was including the Shared items project, which as you suggested, I'll be reverting.

File: vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp (hunk 3 above)
Comment (JunielKatarn) [reply vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:71]: Error prone indeed. I was getting crashes during my testing.

I'll add a comment to explain why this is done.

File: vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp [hunk 7]
```diff
@@ -35,6 +40,11 @@ bool HasPackageIdentity() noexcept {
   return hasPackageIdentity;
 }
 
+ReactPropertyId<bool> HttpUseMonolithicModuleProperty() noexcept {
```
Comment (JunielKatarn) [inline vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:43]: @asklar Please review this usage of the React Context's property bag.

File: vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp [hunk 8]
```diff
@@ -35,6 +39,11 @@ bool HasPackageIdentity() noexcept {
   return hasPackageIdentity;
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:39]: just noticing now that this functionality is duplicated in 3 other places:

- https://github.com/microsoft/react-native-windows/blob/0266f65ffeb0eef353612f9fedb89e3d5bfc6abb/vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp#L473
- https://github.com/microsoft/react-native-windows/blob/main/vnext/Shared/DevSupportManager.cpp#L252
- https://github.com/microsoft/react-native-windows/blob/main/vnext/Shared/Utils.cpp#L60

Consider consolidating into a shared location

File: vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp (hunk 7 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:42]: We've used the convention of namespace="ReactNative.xyz", prop name="Something".
I see that there are a few modules that diverged from this:
- http module
- blob module
- web socket module

all the other ones follow the existing convention. It would be good to stick to the convention for this module (it's fine to fix up the other ones in a separate PR).

So for this one I think we'd say:
```cpp
  static ReactPropertyId<bool> propId{L"ReactNative.HttpModul", L"UseMonolithicModule"};
```

similar in the other ones

Comment (asklar) [reply vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:42]: 1) also I'm not super clear on what UseMonolithicModule means, consider a more descriptive name.

2) Is the idea that CreateHttpModule/"Networking" provides blob/filereader, or equivalent functionality, and then we override it if the "UseMonolithicModule" property isn't set?

3) under which conditions would you choose to use UseMonolithicModule or not? is there a tradeoff/benefit of setting it?

Comment (JunielKatarn) [reply vnext/Microsoft.ReactNative/Base/CoreNativeModules.cpp:42]: The "monolithic module" is just the previous HTTP module implementation, which shoved all the business logic into the CxxModule subclass.

We are still giving an option to use that module in case of any issues with the current one, but we'll be most likely just deleting it soon.
This older module is not compatible with Blob support.

**Comment by JunielKatarn** [comment]
Ready to merge.
